openff/toolkit/utils/rdkit_wrapper.py

```diff
@@ -334,10 +389,231 @@ def _polymer_openmm_pdbfile_to_offtop(
+                mol_smarts = Chem.MolToSmarts(qmol)
+                raise SubstructureAtomSmartsInvalid(
+                    name, atom.GetSmarts(), mol_smarts, error_reason
+                )
+            if not atom.Match(atom):
+                error_reason = f"query does not match rdchem.Mol reading of the molecule (likely due to incorrect/ambiguous connectivity)"
+                mol_smarts = Chem.MolToSmarts(qmol)
+                raise SubstructureAtomSmartsInvalid(
+                    name, atom.GetSmarts(), mol_smarts, error_reason
+                )
+            return True
+
+        def is_valid_neighbor_atom(atom, qmol):
+            atom_smarts = atom.GetSmarts()
+            # ensure that no unsupported logical operators exist
+            operators = r"!,;&"
+            found_ops = [op in atom_smarts for op in operators]
+            if any(found_ops):
+                operator_chars = [op for op, b in zip(operators, found_ops) if b]
+                error_reason = (
+                    f"found unsupported logical operator(s): {operator_chars}"
+                )
+                mol_smarts = Chem.MolToSmarts(qmol)
+                raise SubstructureAtomSmartsInvalid(
+                    name, atom.GetSmarts(), mol_smarts, error_reason
+                )
+            # ensure that no other unsupported atomic primitives are accepted
+            unsupported_prims = r"@xXDvrRhH"
+            found_prims = [prim in atom_smarts for prim in unsupported_prims]
+            if any(found_prims):
+                operator_chars = [
+                    prim for prim, b in zip(unsupported_prims, found_prims) if b
+                ]
+                error_reason = f"found unsupported primitive(s): {operator_chars}"
+                mol_smarts = Chem.MolToSmarts(qmol)
+                raise SubstructureAtomSmartsInvalid(
+                    name, atom.GetSmarts(), mol_smarts, error_reason
+                )
+            # require that atoms have a wildtype atom and label
+            required_prims = r"[]*:"
+            missing_prims = [prim not in atom_smarts for prim in required_prims]
+            if any(missing_prims):
+                operator_chars = [
+                    prim for prim, b in zip(required_prims, missing_prims) if b
+                ]
+                error_reason = f"required primitive(s) not included: {operator_chars}"
+                mol_smarts = Chem.MolToSmarts(qmol)
+                raise SubstructureAtomSmartsInvalid(
+                    name, atom.GetSmarts(), mol_smarts, error_reason
+                )
+            return True
+
+        def is_valid_bond(bond):
+            valid_bond_types = [
+                Chem.BondType.SINGLE,
+                Chem.BondType.AROMATIC,
+                Chem.BondType.DOUBLE,
+                Chem.BondType.TRIPLE,
+                Chem.BondType.QUADRUPLE,
+                Chem.BondType.QUINTUPLE,
```

Comment:
The main issue with aromatic groups is that they can trick the code later into thinking that those substructures give chemically ambiguous assignments. In aromatic groups, the graph match will often return matches that maybe flip or rotate aromatic rings, so when a single bond is overwritten as a double, the expected behavior is to error. While the `_get_symmetrical_groups` function helps somewhat with aromatic groups, I think there may be edge cases that require the use of aromatic bonds (I think polyphenylene with all its aromatic rings is a good example of such a case). 

Could we make it work if we require (and somehow enforce) the use of the MDL aromaticity model on the substructures?

Reply:
I see your point and am thinking about it. Supporting aromatic bonds is unsatisfying/dangerous on a fundamental level - the information content of a graph molecule only needs two of (bond order, formal charge, aromaticity). If we have all three, the molecule is overdefined, and the sources of information become able to disagree with each other, leading to unnecessary edge cases that wouldn’t be there if only two of them were known.

What about requiring the “stable” (on disk) representations to be kekulized (so only integer bond order+formal charge), but then perceiving MDL aromaticity upon loading and during matching, and kekulizing the final loaded polymer at the end? Then we should get all the benefits of aromaticity in avoiding matching conflicts, but avoid using aromaticity in our input or outputs.

Reply:
The "two degrees of freedom" argument makes a ton of sense actually, and I now agree aromaticity could be an issue. I will try out that solution that you mentioned. It should work since all of the aromaticity seen in my/Tim's examples use simple/MDL aromaticity. Thanks!